tox_docker/tox4/config.py

```diff
@@ -0,0 +1,189 @@
+from collections import defaultdict
+from typing import Callable, Container, Dict, List, Sequence
+
+from tox.config.loader.section import Section
+from tox.config.main import Config
+from tox.config.sets import ConfigSet
+from tox.tox_env.api import ToxEnv
+
+from tox_docker.config import (
+    ContainerConfig,
+    RunningContainers,
+    validate_link,
+    validate_port,
+    validate_volume,
+)
+
+# nanoseconds in a second; named "SECOND" so that "1.5 * SECOND" makes sense
+SECOND = 1000000000
+
+EnvRunningContainers = Dict[ToxEnv, RunningContainers]
+
+
+class MissingRequiredSetting(Exception):
+    pass
+
+
+def required(setting_name: str) -> Callable[[str], str]:
+    def require_value(val: str) -> str:
+        if not val:
+            raise MissingRequiredSetting(setting_name)
+        return val
+
+    return require_value
+
+
+class EnvDockerConfigSet(ConfigSet):
+    def register_config(self) -> None:
+        self.add_config(
+            keys=["docker"],
+            of_type=List[str],
+            default=[],
+            desc="docker image configs to load",
+        )
+
+
+class DockerConfigSet(ConfigSet):
+    def register_config(self) -> None:
+        self.add_config(
+            keys=["image"],
+            of_type=str,
+            default="",
```

Comment:
empty? shouldn't be None if not set?

Reply:
What's the right way to make it "required"? I use a post-processor to accomplish that, but that feels a bit janky...

Reply:
Nothing in tox is required, but instead, on ``__init__`` might raise as ValueError to mark it in essence required.